### PR TITLE
chore(gatsby): convert babel-loader-helpers to typescript

### PR DESCRIPTION
**Pull request by Kornil**
## Description

Convert babel-loader-helpers to TS.

## Related Issues

Previous PR was forgotten and recently closed, as suggested in the closing message, I updated the code and re-opened it. https://github.com/gatsbyjs/gatsby/pull/25100
- https://github.com/gatsbyjs/gatsby/issues/21995

### CODE REVIEW

**Comment by Kornil** [comment]
Hi @j0sh77, I've seen the e2e tests failing so I re-merged the master branch, but I see the tests not starting again. Should I just re-cherry-pick your commits to kick off the build?
I'm not sure how to proceed from here.

**Comment by j0sh77** [comment]
> Hi @j0sh77, I've seen the e2e tests failing so I re-merged the master branch, but I see the tests not starting again. Should I just re-cherry-pick your commits to kick off the build? I'm not sure how to proceed from here.

Yeah, there seems to be an issue with builds kicking off commits from forked repos? 

The production runtime e2e tests are flaky, and the react 18 ones are perpetual failures at this point. We're hoping to get those addressed soon, but you can ignore them for the time being.

Again, thank you so much for this PR and taking the time to rebase. 💜

**Comment by Kornil** [comment]
Thank you for your answer, that explained a lot!

I see now only one react_18 test is failing (remote-file.js).
I've checked it but does not seem to have any relation to these changes so I'll just assume it's a flake.

File: packages/gatsby/src/utils/babel-loader-helpers.ts [hunk 1]
```diff
@@ -136,24 +154,27 @@ const prepareOptions = (
...
-    )
-  })
+  const reduxPlugins: Array<PluginItem> = []
+  const reduxPresets: Array<PluginItem> = []
+
+  if (stage) {
+    pluginBabelConfig.stages[stage].plugins.forEach(plugin => {
+      reduxPlugins.push(
+        babel.createConfigItem([resolve(plugin.name), plugin.options], {
+          dirname: plugin.name,
```
Comment (j0sh77) [inline packages/gatsby/src/utils/babel-loader-helpers.ts:164]: I think adding the `dirname` is ok here, but should we keep the `name` key as well?

https://github.com/babel/babel/blob/main/packages/babel-core/src/config/item.ts

Comment (Kornil) [reply packages/gatsby/src/utils/babel-loader-helpers.ts:164]: I was going to leave it but the types from babel changed and no longer include it.
The `createConfigItem` method that we use here comes from this type https://github.com/babel/babel/blob/main/packages/babel-core/src/config/item.ts#L28 and it now accepts only `dirname` and `type`.

Comment (LekoArts) [reply packages/gatsby/src/utils/babel-loader-helpers.ts:164]: Looking at the current babel docs and previous versions the createConfigItem now has these parameters so changing them feels correct

**Review by j0sh77 — APPROVED** [review]
Thanks again for your patience, understanding, and contribution to Gatsby! 💜

**Pull request merged**